chapters/untyped_pointers.md

```diff
@@ -0,0 +1,129 @@
+
+# Rethinking the pointer type
+
+The better understand this extension, first focus on `OpTypeUntypedPointerKHR` as that is the "core" change. By using this new pointer type, you will find the reason the other "untyped" instructions were added to go along with it.
+
+Let's use this very basic example of load and storing to a storage buffer ([view online](https://godbolt.org/z/5Wx1PsxG7)):
+
+```glsl
+layout(set = 0, binding = 0) buffer SSBO {
+    uint a;
+};
+
+void main() {
+    uint b = a;
+    a = 0;
+}
+```
+
+This will generate into something such as
+
+```swift
+   %uint = OpTypeInt 32 0
+ %uint_0 = OpConstant %uint 0
+
+    %SSBO = OpTypeStruct %uint
+%SSBO_ptr = OpTypePointer StorageBuffer %SSBO
+%variable = OpVariable %SSBO_ptr StorageBuffer
+
+ %ptr = OpTypePointer StorageBuffer %uint
+
+  %ac = OpAccessChain %ptr %variable %uint_0
+%load = OpLoad %uint %ac
+        OpStore %ac %uint_0
+```
+
+Notice that `OpLoad` and `OpStore` both have two operands, and **both** can figure out the type is `%uint`. This is the **core motivation** of this extension, to remove the need for having to declare the `%uint` type here in the `Pointer` operand.
+
+With untyped pointer, we still want will use the access chain to get the `StorageClass`, but we can now remove the redundancy by first turning the pointer into an untyped pointer.
+
+```diff
+   %uint = OpTypeInt 32 0
+ %uint_0 = OpConstant %uint 0
+
+- %ptr = OpTypePointer StorageBuffer %uint
++ %ptr = OpTypeUntypedPointerKHR StorageBuffer
+
+    %SSBO = OpTypeStruct %uint
+%SSBO_ptr = OpTypePointer StorageBuffer %SSBO
+%variable = OpVariable %SSBO_ptr StorageBuffer
+
+  %ac = OpAccessChain %ptr %variable %uint_0
+%load = OpLoad %uint %ac
+        OpStore %ac %uint_0
+```
+
+But this is **stil invalid**! We now need a matching untyped access chain as well where we now have a new `Base Type` operand.
+
+```diff
+   %uint = OpTypeInt 32 0
+ %uint_0 = OpConstant %uint 0
+
+ %ptr = OpTypeUntypedPointerKHR StorageBuffer
+
+    %SSBO = OpTypeStruct %uint
+%SSBO_ptr = OpTypePointer StorageBuffer %SSBO
+%variable = OpVariable %SSBO_ptr StorageBuffer
+
+-  %ac = OpAccessChain %ptr %variable %uint_0
++  %ac = OpUntypedAccessChainKHR %ptr %SSBO %variable %uint_0
+%load = OpLoad %uint %ac
+        OpStore %ac %uint_0
+```
+
+From here, we can now go an extra step and turn the `OpVariable` into an untyped variable as it now doesn't requires to be tied to a pointer anymore.
+
+This is the same mental model of taking a `ssbo* ptr` and now making it `void* ptr`
+
+```diff
+   %uint = OpTypeInt 32 0
+ %uint_0 = OpConstant %uint 0
+
+ %ptr = OpTypeUntypedPointerKHR StorageBuffer
+
+    %SSBO = OpTypeStruct %uint
+- %SSBO_ptr = OpTypePointer StorageBuffer %SSBO
+- %variable = OpVariable %SSBO_ptr StorageBuffer
++ %variable = OpUntypedVariableKHR %ptr StorageBuffer %SSBO
```

Comment:
Maybe worth pointing out that this is optional. To facilitate transitions, untyped pointer opcodes generally support the input pointers being typed or untyped and generate an untyped result.